### ISSUE TITLE
Move code to cross-platform C++...

### ISSUE THREAD

**Issue by FredM67** [pistacheio/pistache#975]
The doc says "It is entirely written in pure-C++17"... why is the code not 100% pure C++17 ?
event_fd is linux specific for example. That could be replaced with condition variable and Co.

With 100% pure C++17, the code could be fully cross-platform and so will gain support for Windows and MacOs.

**Comment by kiplingw** [comment]
Thanks @FredM67. If you'd like to submit a PR and with corresponding unit tests, one of us would be happy to take a look.

**Comment by FredM67** [comment]
Well, first I've to find time for that ! I'll try...

**Comment by Tachi107** [comment]
Hi @FredM67, as you probably know C++ doesn't have networking support, so it is impossible to implement a REST framework without using system-specific APIs.
Pistache is designed around Linux's `epoll`, and other platforms don't have a similar API (see details in https://github.com/pistacheio/pistache/issues/6#issuecomment-242398225).
It may be possible to switch to `io_uring` a newer Linux API that seems to have a Windows counterpart, and should also be faster.

**Comment by FredM67** [comment]
Oh yes sorry, I did forget the network stuff....
Indeed, all the thread sync, notifications, ... could be achieved with cross-platform C++ with the STL.

**Comment by TabrynP** [comment]
Hello! I don't have a ton of time on my hands, but assuming time is not of the essence I think I might be able to assist. I have a good deal of experience in [Boost Asio](https://www.boost.org/doc/libs/1_76_0/doc/html/boost_asio.html) as well as the asynchronous programming features of the standard library (`std::async`, `std::promise/std::future`, etc), Linux socket networking, and user space networking.

First question about a potential cross-platform network implementation: is the Reactor pattern a strict requirement? If yes, I would be curious to know why! Not planning on pushing back, just for curiosity's sake.

**Comment by kiplingw** [comment]
@TabrynP, removing the reactor pattern would be a major pain. It's the core design pattern basically everything important is built around. 

Having said that, the reactor pattern itself doesn't require the BSD sockets interface we are using, _per se_. But if the network API itself was replaced, it would make sense to do so with the upcoming STL [networking API](https://en.cppreference.com/w/cpp/experimental/networking) as the way to go. It's based on the Boost Asio library's design.

**Comment by TabrynP** [comment]
@kiplingw Makes sense to me! It seems like the latest versions of both Boost and non-boost Asio implement the new networking API. Are either of those libraries acceptable dependencies? That would be the shortest path. 

Re-implementing the existing networking features according to the standard API and adding cross-platform functionality along the way is also reasonable, but it would take considerably longer.

**Comment by Tachi107** [comment]
I originally chose Pistache mainly because of its performance, but I don't know if Asio will be able to deliver the same speed; it should be important to take a look at some benchmarks, maybe also taking a look at Boost.Beast

**Comment by kiplingw** [comment]
I think adding a new dependency isn't really the issue (for me at least). That's something the package manager solves when the user pulls the library. It's a question of which dependency. I think the best thing to do is wait until the STL standardizes the networking API and implementations become ubiquitous.
